docs/proposals/20220712-cluster-api-addon-orchestration.md

```diff
@@ -0,0 +1,448 @@
+	// +optional
+	Conditions clusterv1.Conditions `json:"conditions,omitempty"`
+
+	// Status is the current status of the Helm release.
+	// +optional
+	Status string `json:"status,omitempty"`
+
+	// Revision is the current revision of the Helm release.
+	// +optional
+	Revision int `json:"revision,omitempty"`
+
+	// Namespace is the namespace of the Helm release on the workload cluster.
+	// +optional
+	Namespace string `json:"namespace,omitempty"`
+}
+```
+
+#### Controller Design
+
+The HelmChartProxy controller is responsible for maintaining our list of HelmReleaseProxies and resolving any referenced Cluster fields. The controller will:
+
+- Find all workload clusters matching the cluster selector.
+- Resolve the referenced Cluster fields in the values map based on each selected cluster.
+- Create a HelmReleaseProxy for each selected workload cluster if one does not exist with the resolved values.
+- Update the HelmReleaseProxy for each selected workload cluster if it exists and the version or resolved values have changed.
+- Delete the HelmReleaseProxy for each selected workload cluster if the cluster no longer matches the cluster selector.
+
+The HelmReleaseProxy controller is responsible for creating, updating, and deleting the Helm release installed on the workload cluster. The controller will:
+
+- Install a Helm release on the workload cluster if it does not exist.
+- Update the existing Helm release if the HelmReleaseProxy values or version do not match the existing Helm release.
+- Delete the Helm release if the HelmReleaseProxy is deleted.
+
+Note that the HelmReleaseProxy controller contains a static configuration and is only responsible for ensuring there is a Helm release matching the HelmReleaseProxy. The HelmReleaseProxy gets created, updated, and deleted by the HelmChartProxy controller.
+
+#### Considerations about the CRD design
+
+The scope of the cluster values accessible using Go templating must be defined along with use cases. An option would be to expose the Builtin struct type used in ClusterClass or construct a similar type. Additionally, while resources like the control plane and cluster have a 1:1 correlation with each selected cluster, resources like MachineDeployments or Machines could have any number and can’t be selected in a similar way. As a result, additional work is needed to decide if there’s a use case for including information in those types.
+
+### Example: Installing Calico CNI
+
+Developers and Cluster operators often want to install a CNI on their workload clusters. HelmChartProxy controller can be used to install the Calico CNI on each selected workload cluster. In this example, the value configuration can be leveraged to initialize the `installation.calicoNetwork.ipPools` array by creating an element for every pod CIDR in the Cluster.
+
+```yaml
+apiVersion: addons.cluster.x-k8s.io/v1beta1
+kind: HelmChartProxy
+metadata:
+  name: calico-cni
+spec:
+  clusterSelector:
```

Comment:
does this apply to all namespaces?

Reply:
At the moment it does, though I've been thinking about adding a `namespaces` or `allowedNamespaces` field to the selector.